# Penerapan Metode AHP dan SAW dalam Pemilihan Hero Mobile Legend pada MPL Season 13

Dataset source: https://www.kaggle.com/datasets/bcakra/tournamnet-result-match-of-mpl-indonesia-season-13?select=MPL+Indonesia+Season+13+-+Hero_Role.csv

## Data Wrangling

### Loading Data

In [86]:
# Import libraries
import pandas as pd
import numpy as np

In [87]:
# Load the data
data = pd.read_csv('data/MPL Indonesia Season 13 - Hero_Role.csv')
data.head()

Hero                       Title  Hero Order    Role(s) Specialty(ies)  \
0     Miya  Miya, the Moonlight Archer           1   Marksman       Finisher   
1  Balmond   Balmond, the Bloody Beast           2    Fighter         Damage   
2    Saber  Saber, the Wandering Sword           3   Assassin         Charge   
3    Alice   Alice, the Queen of Blood           4  Mage/Tank   Charge/Regen   
4     Nana      Nana, the Sweet Leonin           5       Mage           Poke   

  Lane Recommendation(s)  Health  Health Regen   Mana  Mana Regen  ...  \
0              Gold Lane    2524           6.0  445.0        15.0  ...   
1      Jungling/EXP Lane    2736           9.4    0.0         0.0  ...   
2               Jungling    2599           7.0  443.0        16.0  ...   
3      EXP Lane/Mid Lane    2573           7.2  493.0        18.0  ...   
4       Mid Lane/Roaming    2501           6.8  510.0        18.0  ...   

                                    Pasive Deskripsi  Skill 1 Name Abilities  \
0  Miya gains 5% extra Attack Speed for 4s each t...              Moon Arrow   
1  Balmond recovers 5% Max HP on minion or creep ...               Soul Lock   
2  Saber's attacks reduce enemies' Physical Defen...         Orbiting Swords   
3  A blood orb will be generated each time an ene...           Flowing Blood   
4  Upon taking fatal damage, Nana removes all deb...         Magic Boomerang   

   Skill 1 Ability                           Skill 1 Pasive Deskripsi  \
0         BUFF AoE  Miya shoots two extra arrows on each Basic Att...   
1       Blink Slow  Balmond charges in the target direction for a ...   
2         AoE BUFF  Saber releases 5 swords that orbit around him ...   
3        Blink AoE  Alice casts a sphere of blood energy in the ta...   
4              AoE  Nana hurls her Magic Boomerang in the target d...   

   Skill 2 Name Abilities  Skill 2 Ability  \
0        Arrow of Eclipse           CC AoE   
1           Cyclone Sweep         AoE BUFF   
2                  Charge        Blink AoE   
3               Blood Awe           CC AoE   
4           Molina Smooch        CC Debuff   

                            Skill 2 Pasive Deskripsi  Skill 3 Name Abilities  \
0  Miya launches an empowered arrow on the target...        Hidden Moonlight   
1  Balmond spins with his axe for 3s, continuousl...          Lethal Counter   
2  Saber dashes in the target direction, dealing ...            Triple Sweep   
3  Alice deals 270(+90%Total Magic Power) Magic D...               Blood Ode   
4  Nana sends Molina to the target location, who ...            Molina Blitz   

     Skill 3 Ability                           Skill 3 Pasive Deskripsi  
0  Conceal Remove CC  Miya removes all debuffs on her and conceals h...  
1         Burst Slow  After a short delay, Balmond smashes his axe i...  
2           Burst CC  Saber charges at the target enemy hero, knocki...  
3           AoE HEAL  Alice enters the blood-draining state, consumi...  
4           Burst CC  Nana summons Molina to attack the target area ...  

[5 rows x 34 columns]

### Assessing Data

In [88]:
# Check the shape of the data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 34 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Hero                      123 non-null    object 
 1   Title                     123 non-null    object 
 2   Hero Order                123 non-null    int64  
 3   Role(s)                   123 non-null    object 
 4   Specialty(ies)            123 non-null    object 
 5   Lane Recommendation(s)    123 non-null    object 
 6   Health                    123 non-null    int64  
 7   Health Regen              123 non-null    float64
 8   Mana                      118 non-null    float64
 9   Mana Regen                118 non-null    float64
 10  Attack Speed              123 non-null    object 
 11  Physical Damage           123 non-null    int64  
 12  Magical Damage            123 non-null    int64  
 13  Physical Defense          123 non-null    int64  
 14  Magical De

Total 34 kolom dan 123 entri hero. Terdapat beberapa tipe data pada kolom krusial yang belum sesuai, seperti win rate, price, dan attack speed.

In [89]:
data.nunique()

Hero                        123
Title                       123
Hero Order                  123
Role(s)                      24
Specialty(ies)               40
Lane Recommendation(s)       18
Health                       83
Health Regen                 26
Mana                         43
Mana Regen                   19
Attack Speed                 35
Physical Damage              31
Magical Damage                1
Physical Defense             16
Magical Defense               2
Movement Speed               13
Energy                        1
Energy Regen                  5
Price                        13
Specialty                    61
Win Rate                    123
Hero.1                      123
 Name Pasive Abilities      123
Ability                      30
Pasive Deskripsi            123
Skill 1 Name Abilities      123
Skill 1 Ability              41
Skill 1 Pasive Deskripsi    123
Skill 2 Name Abilities      123
Skill 2 Ability              59
Skill 2 Pasive Deskripsi    123
Skill 3 

In [90]:
# Mengecek nilai null
data.isna().sum()

Hero                          0
Title                         0
Hero Order                    0
Role(s)                       0
Specialty(ies)                0
Lane Recommendation(s)        0
Health                        0
Health Regen                  0
Mana                          5
Mana Regen                    5
Attack Speed                  0
Physical Damage               0
Magical Damage                0
Physical Defense              0
Magical Defense               0
Movement Speed                0
Energy                      117
Energy Regen                117
Price                         1
Specialty                     3
Win Rate                      0
Hero.1                        0
 Name Pasive Abilities        0
Ability                       0
Pasive Deskripsi              0
Skill 1 Name Abilities        0
Skill 1 Ability               0
Skill 1 Pasive Deskripsi      0
Skill 2 Name Abilities        0
Skill 2 Ability               0
Skill 2 Pasive Deskripsi      0
Skill 3 

Terdapat 5 entri hero yang tidak menggunakan mana/mana regen melainkan energi/regen. 1 entri hero juga yang pricenya null. Untuk kolom specialty, dikarenakan tidak relevant dengan tujuan yg dicapai maka abaikan saja

In [91]:
data.describe()

Hero Order       Health  Health Regen        Mana  Mana Regen  \
count  123.000000   123.000000    123.000000  118.000000  118.000000   
mean    62.000000  2580.544715      7.681301  347.211864   12.708475   
std     35.651087   207.092529      1.592161  214.532322    8.567745   
min      1.000000   918.000000      3.800000    0.000000    0.000000   
25%     31.500000  2501.000000      6.800000    0.000000    0.000000   
50%     62.000000  2578.000000      7.200000  440.000000   15.000000   
75%     92.500000  2701.000000      8.000000  490.000000   18.000000   
max    123.000000  2909.000000     18.400000  750.000000   31.000000   

       Physical Damage  Magical Damage  Physical Defense  Magical Defense  \
count       123.000000           123.0        123.000000       123.000000   
mean        116.065041             0.0         19.154472        14.878049   
std           9.835316             0.0          3.550646         0.774442   
min          90.000000             0.0         10.000000        10.000000   
25%         110.000000             0.0         17.000000        15.000000   
50%         117.000000             0.0         19.000000        15.000000   
75%         122.000000             0.0         21.500000        15.000000   
max         140.000000             0.0         27.000000        15.000000   

       Movement Speed  Energy  Energy Regen  
count      123.000000     6.0      6.000000  
mean       251.495935   100.0     42.800000  
std          8.715328     0.0     96.618425  
min        240.000000   100.0      0.800000  
25%        240.000000   100.0      3.250000  
50%        252.000000   100.0      4.000000  
75%        260.000000   100.0      4.750000  
max        270.000000   100.0    240.000000

- Dikarenakan kolom magical damage bernilai 0 semua dan hero magic pun mempunyai physical damage pada tabel, anggap physical damage sebagai damage universal.
- Dikarenakan beberapa hero ada yang tidak menggunakan mana sama sekali dan beberapa hero juga menggunakan energi. Penggunaan mana,energi,mana regen, dan energi regen diabaikan saja dari kriteria.

In [92]:
print("Jumlah duplikasi: ", data.duplicated().sum())

Jumlah duplikasi:  0


### Data Cleaning

In [93]:
clean_data = data.copy()

#### Menyesuaikan tipe data

Mengubah tipe data win rate, price, dan attack speed menjadi float

In [94]:
# Mengecek tipe data win rate, price, dan attack speed
clean_data[['Win Rate', 'Price', 'Attack Speed']].dtypes

Win Rate        object
Price           object
Attack Speed    object
dtype: object

In [95]:
# Melihat entri data win rate, price, dan attack speed
clean_data[['Attack Speed', 'Price', 'Win Rate']].head()

Attack Speed        Price              Win Rate
0  1.06 ┃ 100%        10800    33W : 32L (50.77%)
1  1.06 ┃ 100%   6500   299  453W : 516L (46.75%)
2  1.08 ┃ 100%   6500   299  142W : 146L (49.31%)
3     1 ┃ 100%  15000   399  513W : 526L (49.37%)
4  1.08 ┃ 100%   6500   299   74W : 105L (41.34%)

In [96]:
# Mengganti nilai data attack speed
clean_data['Attack Speed'] = clean_data['Attack Speed'].str.split('┃').str[0].astype(float)

In [97]:
# Mengganti nilai data win rate
clean_data['Win Rate'] = clean_data['Win Rate'].str.extract(r'(\d+(\.\d+)?)%\)$')[0].astype(float)

In [98]:
# Mengganti nilai data price
clean_data['Price'] = clean_data['Price'].str.split().str[0].astype(float)

In [99]:
# Melihat entri data win rate, price, dan attack speed
clean_data[['Attack Speed', 'Price', 'Win Rate']].head()

Attack Speed    Price  Win Rate
0          1.06  10800.0     50.77
1          1.06   6500.0     46.75
2          1.08   6500.0     49.31
3          1.00  15000.0     49.37
4          1.08   6500.0     41.34

In [100]:
# Mengecek Tipe Data
clean_data[['Win Rate', 'Price', 'Attack Speed']].dtypes

Win Rate        float64
Price           float64
Attack Speed    float64
dtype: object

#### Mengisi nilai price yang hilang

In [101]:
# Melihat hero yang mempunyai price yang hilang.
clean_data[clean_data['Price'].isna()]

Hero                    Title  Hero Order Role(s) Specialty(ies)  \
118  Novaria  Novaria, the Star Rebel         119    Mage          Burst   

    Lane Recommendation(s)  Health  Health Regen   Mana  Mana Regen  ...  \
118               Mid Lane    2401           6.8  540.0         4.6  ...   

                                      Pasive Deskripsi  \
118  Astral Spheres summoned by Novaria will contin...   

     Skill 1 Name Abilities  Skill 1 Ability  \
118           Astral Meteor         Slow AoE   

                              Skill 1 Pasive Deskripsi  \
118  Novaria summons an Astral Sphere at the target...   

     Skill 2 Name Abilities  Skill 2 Ability  \
118           Astral Recall       Burst BUFF   

                              Skill 2 Pasive Deskripsi  \
118  Novaria summons an Astral Sphere at a distance...   

     Skill 3 Name Abilities  Skill 3 Ability  \
118             Astral Echo             BUFF   

                              Skill 3 Pasive Deskripsi  
118  Novaria scatters Astral Echo in the target dir...  

[1 rows x 34 columns]

Hero yang mempunyai price yang hilang adalah hero yang bernama "Novaria". Isi nilai yang hilang dengan 32000 karena harga hero novaria di game adalah 32000.

In [102]:
# Mengisi nilai
clean_data['Price'].fillna(32000, inplace=True)

C:\Users\ASUS\AppData\Local\Temp\ipykernel_20396\2081978115.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  clean_data['Price'].fillna(32000, inplace=True)


In [103]:
# Mengecek nilai null pada price kembali
clean_data['Price'].isna().sum()

0

In [104]:
# Mengecek price hero novaria
clean_data[clean_data['Hero'] == 'Novaria'].Price

118    32000.0
Name: Price, dtype: float64

#### Menyeleksi Fitur yang Dibutuhkan

In [105]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 34 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Hero                      123 non-null    object 
 1   Title                     123 non-null    object 
 2   Hero Order                123 non-null    int64  
 3   Role(s)                   123 non-null    object 
 4   Specialty(ies)            123 non-null    object 
 5   Lane Recommendation(s)    123 non-null    object 
 6   Health                    123 non-null    int64  
 7   Health Regen              123 non-null    float64
 8   Mana                      118 non-null    float64
 9   Mana Regen                118 non-null    float64
 10  Attack Speed              123 non-null    float64
 11  Physical Damage           123 non-null    int64  
 12  Magical Damage            123 non-null    int64  
 13  Physical Defense          123 non-null    int64  
 14  Magical De

In [106]:
features_needed = ['Hero', 'Health', 'Health Regen', 'Attack Speed', 'Physical Damage', 'Physical Defense', 'Magical Defense', 'Movement Speed', 'Price', 'Win Rate']

In [107]:
final_data = clean_data[features_needed].copy()

In [108]:
# Rename physical damage column to damage
final_data.rename(columns={'Physical Damage': 'Damage'}, inplace=True)

In [109]:
final_data.head()

Hero  Health  Health Regen  Attack Speed  Damage  Physical Defense  \
0     Miya    2524           6.0          1.06     100                17   
1  Balmond    2736           9.4          1.06     119                25   
2    Saber    2599           7.0          1.08     118                20   
3    Alice    2573           7.2          1.00     114                21   
4     Nana    2501           6.8          1.08     115                17   

   Magical Defense  Movement Speed    Price  Win Rate  
0               15             240  10800.0     50.77  
1               15             260   6500.0     46.75  
2               15             260   6500.0     49.31  
3               15             240  15000.0     49.37  
4               15             250   6500.0     41.34

In [110]:
final_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Hero              123 non-null    object 
 1   Health            123 non-null    int64  
 2   Health Regen      123 non-null    float64
 3   Attack Speed      123 non-null    float64
 4   Damage            123 non-null    int64  
 5   Physical Defense  123 non-null    int64  
 6   Magical Defense   123 non-null    int64  
 7   Movement Speed    123 non-null    int64  
 8   Price             123 non-null    float64
 9   Win Rate          123 non-null    float64
dtypes: float64(4), int64(5), object(1)
memory usage: 9.7+ KB


In [111]:
final_data.describe()

Health  Health Regen  Attack Speed      Damage  Physical Defense  \
count   123.000000    123.000000    123.000000  123.000000        123.000000   
mean   2580.544715      7.681301      1.040650  116.065041         19.154472   
std     207.092529      1.592161      0.059736    9.835316          3.550646   
min     918.000000      3.800000      0.870000   90.000000         10.000000   
25%    2501.000000      6.800000      1.000000  110.000000         17.000000   
50%    2578.000000      7.200000      1.030000  117.000000         19.000000   
75%    2701.000000      8.000000      1.080000  122.000000         21.500000   
max    2909.000000     18.400000      1.220000  140.000000         27.000000   

       Magical Defense  Movement Speed         Price   Win Rate  
count       123.000000      123.000000    123.000000  123.00000  
mean         14.878049      251.495935  28222.756098   48.74813  
std           0.774442        8.715328   8120.532319    4.76978  
min          10.000000      240.000000    599.000000   25.00000  
25%          15.000000      240.000000  32000.000000   46.95500  
50%          15.000000      252.000000  32000.000000   49.37000  
75%          15.000000      260.000000  32000.000000   51.10500  
max          15.000000      270.000000  32000.000000   61.11000

In [112]:
# Simpan data yang sudah final
final_data.to_csv('./data/final_data.csv', index=False)

# Processing Data

In [113]:
df = pd.read_csv("data/final_data.csv")

In [114]:
df.head()

Hero  Health  Health Regen  Attack Speed  Damage  Physical Defense  \
0     Miya    2524           6.0          1.06     100                17   
1  Balmond    2736           9.4          1.06     119                25   
2    Saber    2599           7.0          1.08     118                20   
3    Alice    2573           7.2          1.00     114                21   
4     Nana    2501           6.8          1.08     115                17   

   Magical Defense  Movement Speed    Price  Win Rate  
0               15             240  10800.0     50.77  
1               15             260   6500.0     46.75  
2               15             260   6500.0     49.31  
3               15             240  15000.0     49.37  
4               15             250   6500.0     41.34

In [115]:
# Select the columns from "Health" to "Win Rate"
columns_of_interest = df.loc[:, 'Health':'Win Rate']

# Convert to a NumPy array
array = columns_of_interest.to_numpy()

# Set print options to suppress scientific notation
np.set_printoptions(suppress=True)

# Print the resulting array
print(array)

[[ 2524.       6.       1.06 ...   240.   10800.      50.77]
 [ 2736.       9.4      1.06 ...   260.    6500.      46.75]
 [ 2599.       7.       1.08 ...   260.    6500.      49.31]
 ...
 [ 2620.       8.       1.04 ...   240.   32000.      54.96]
 [ 2549.       7.       1.08 ...   260.   32000.      61.11]
 [ 2609.       6.       1.07 ...   260.   32000.      58.93]]


In [116]:
# Define matrix comparison
comparison_matrix = np.array([
    [1,   2,   4,   5,   5,   5,   4,   3,   3],
    [1/2, 1,   2,   3,   3,   4,   2,   2,   3],
    [1/4, 1/2, 1,   2,   3,   3,   2,   1,   2],
    [1/5, 1/3, 1/2, 1,   2,   3,   1,   1,   2],
    [1/5, 1/3, 1/3, 1/2, 1,   2,   2,   1,   1],
    [1/5, 1/4, 1/3, 1/3, 1/2, 1,   1,   2,   1],
    [1/4, 1/2, 1/2, 1,   1/2, 1,   1,   1,   2],
    [1/3, 1/2, 1,   1,   1,   1/2, 1,   1,   1],
    [1/3, 1/3, 1/2, 1/2, 1,   1,   1/2, 1,   1]
])

In [117]:
# Sum the normalization matrix
column_sum = comparison_matrix.sum(axis=0)
normalized_matrix = comparison_matrix / column_sum
print(normalized_matrix)

[[0.30612245 0.34782609 0.39344262 0.34883721 0.29411765 0.24390244
  0.27586207 0.23076923 0.1875    ]
 [0.15306122 0.17391304 0.19672131 0.20930233 0.17647059 0.19512195
  0.13793103 0.15384615 0.1875    ]
 [0.07653061 0.08695652 0.09836066 0.13953488 0.17647059 0.14634146
  0.13793103 0.07692308 0.125     ]
 [0.06122449 0.05797101 0.04918033 0.06976744 0.11764706 0.14634146
  0.06896552 0.07692308 0.125     ]
 [0.06122449 0.05797101 0.03278689 0.03488372 0.05882353 0.09756098
  0.13793103 0.07692308 0.0625    ]
 [0.06122449 0.04347826 0.03278689 0.02325581 0.02941176 0.04878049
  0.06896552 0.15384615 0.0625    ]
 [0.07653061 0.08695652 0.04918033 0.06976744 0.02941176 0.04878049
  0.06896552 0.07692308 0.125     ]
 [0.10204082 0.08695652 0.09836066 0.06976744 0.05882353 0.02439024
  0.06896552 0.07692308 0.0625    ]
 [0.10204082 0.05797101 0.04918033 0.03488372 0.05882353 0.04878049
  0.03448276 0.07692308 0.0625    ]]


In [118]:
# Sum each row of the normalization matrix
row_sum = normalized_matrix.sum(axis=1)

# Calculate the priority_scale
priority_scale = row_sum / 9

# Calculate Weighted Sum Vector
weighted_matrix = np.dot(comparison_matrix, priority_scale)

# Calculate Consistency Vector
consistency_vector = weighted_matrix / priority_scale

# Calculate Lambda Max and Consistency Index
lambda_max = consistency_vector.mean()
consistency_index = (lambda_max - 9) / (9 - 1)
print("Lambda max: ", lambda_max)
print("Consistency index (CI): ", consistency_index)


Lambda max:  9.56902111821568
Consistency index (CI):  0.07112763977696002


In [119]:
# Calculate Consistency Ratio
random_index = 1.45 #n=9
consistency_ratio = consistency_index / random_index
print("Consistency ratio (CR): ", consistency_ratio)

Consistency ratio (CR):  0.04905354467376553


### Dikarenakan CR < 0.1, maka comparison_matrix yang berpasangan **konsisten**.

In [120]:
# Define the criteria
criteria = ['Health', 'Health Regen', 'Attack Speed', 'Damage', 'Physical Defense', 'Magical Defense', 'Movement Speed', 'Price', 'Win Rate']
heroes = ['Miya', 'Balmond', 'Saber', 'Alice', 'Nana', 'Tigreal', 'Alucard', 'Karina', 'Akai', 'Franco', 'Bane', 'Bruno', 'Clint', 'Rafaela',
    'Eudora', 'Zilong', 'Fanny', 'Layla', 'Minotaur', 'Lolita', 'Hayabusa', 'Freya', 'Gord', 'Natalia', 'Kagura', 'Chou', 'Sun',
    'Alpha', 'Ruby', 'Yi Sun-shin', 'Moskov', 'Johnson', 'Cyclops', 'Estes', 'Hilda', 'Aurora', 'Lapu-Lapu', 'Vexana', 'Roger',
    'Karrie', 'Gatotkaca', 'Harley', 'Irithel', 'Grock', 'Argus', 'Odette', 'Lancelot', 'Diggie', 'Hylos', 'Zhask', 'Helcurt',
    'Pharsa', 'Lesley', 'Jawhead', 'Angela', 'Gusion', 'Valir', 'Martis', 'Uranus', 'Hanabi', "Chang'e", 'Kaja', 'Selena',
    'Aldous', 'Claude', 'Vale', 'Leomord', 'Lunox', 'Hanzo', 'Belerick', 'Kimmy', 'Thamuz', 'Harith', 'Minsitthar', 'Kadita',
    'Faramis', 'Badang', 'Khufra', 'Granger', 'Guinevere', 'Esmeralda', 'Terizla', 'X.Borg', 'Ling', 'Dyrroth', 'Lylia', 'Baxia', 'Masha',
    'Wanwan', 'Silvanna', 'Cecilion', 'Carmilla', 'Atlas', 'Popol and Kupa', 'Yu Zhong', 'Luo Yi', 'Benedetta', 'Khaleed',
    'Barats', 'Brody', 'Yve', 'Mathilda', 'Paquito', 'Gloo', 'Beatrix', 'Phoveus', 'Natan', 'Aulus', 'Aamon', 'Valentina', 'Edith',
    'Floryn', 'Yin', 'Melissa', 'Xavier', 'Julian', 'Fredrinn', 'Joy','Novaria', 'Arlott', 'Ixia', 'Nolan', 'Cici']

In [121]:
# Define comparison matrices as a list
comparison_matrices = []

# Set the value of matrix evaluation with the criteria
evaluation_matrix = df[criteria].values

# Loop each criteria
for i in range(evaluation_matrix.shape[1]):
    # Get the vector from the matrix evaluation
    criteria_vector = evaluation_matrix[:, i]

    # Create a comparison matrix for this criteria
    comparison_matrix = np.zeros((len(criteria_vector), len(criteria_vector)))
    for j in range(len(criteria_vector)):
        for k in range(len(criteria_vector)):
            comparison_matrix[j, k] = criteria_vector[j] / criteria_vector[k]

    # Add the comparison matrix to the list
    comparison_matrices.append(comparison_matrix)

    # Create the DataFrame from the comparison matrix
    comparison_df = pd.DataFrame(comparison_matrix, columns=heroes, index=heroes)
    print(f"Matriks Perbandingan Berpasangan untuk {criteria[i]}:\n")
    print(comparison_df)
    print("\n")

Matriks Perbandingan Berpasangan untuk Health:

                    Miya   Balmond     Saber     Alice      Nana   Tigreal  \
Miya            1.000000  0.922515  0.971143  0.980956  1.009196  0.938290   
Balmond         1.083994  1.000000  1.052713  1.063350  1.093962  1.017100   
Saber           1.029715  0.949927  1.000000  1.010105  1.039184  0.966171   
Alice           1.019414  0.940424  0.989996  1.000000  1.028788  0.956506   
Nana            0.990887  0.914108  0.962293  0.972017  1.000000  0.929740   
Tigreal         1.065769  0.983187  1.035013  1.045472  1.075570  1.000000   
Alucard         1.038431  0.957968  1.008465  1.018655  1.047981  0.974349   
Karina          1.043185  0.962354  1.013082  1.023319  1.052779  0.978810   
Akai            1.097068  1.012061  1.065410  1.076176  1.107157  1.029368   
Franco          1.073296  0.990132  1.042324  1.052857  1.083167  1.007063   
Bane            1.013867  0.935307  0.984609  0.994559  1.023191  0.951301   
Bruno           

In [122]:
# Calculate a normalization matrix
normalized_matrices = []

for i, comparison_matrix in enumerate(comparison_matrices):
    column_sum = comparison_matrix.sum(axis=0)
    normalized_matrix = comparison_matrix / column_sum
    normalized_matrices.append(normalized_matrix)

    # Create the DataFrame from the normalization matrix
    normalization_df = pd.DataFrame(normalized_matrix, columns=heroes, index=heroes)
    print(f"Matriks Normalisasi {criteria[i]}:\n")
    print(normalization_df)
    print("\n")

Matriks Normalisasi Health:

                    Miya   Balmond     Saber     Alice      Nana   Tigreal  \
Miya            0.007952  0.007952  0.007952  0.007952  0.007952  0.007952   
Balmond         0.008620  0.008620  0.008620  0.008620  0.008620  0.008620   
Saber           0.008188  0.008188  0.008188  0.008188  0.008188  0.008188   
Alice           0.008106  0.008106  0.008106  0.008106  0.008106  0.008106   
Nana            0.007879  0.007879  0.007879  0.007879  0.007879  0.007879   
Tigreal         0.008475  0.008475  0.008475  0.008475  0.008475  0.008475   
Alucard         0.008258  0.008258  0.008258  0.008258  0.008258  0.008258   
Karina          0.008295  0.008295  0.008295  0.008295  0.008295  0.008295   
Akai            0.008724  0.008724  0.008724  0.008724  0.008724  0.008724   
Franco          0.008535  0.008535  0.008535  0.008535  0.008535  0.008535   
Bane            0.008062  0.008062  0.008062  0.008062  0.008062  0.008062   
Bruno           0.007946  0.007946 

In [123]:
# Calculate the priority scale
priority_vectors = []

for i, normalized_matrix in enumerate(normalized_matrices):
    row_sum = normalized_matrix.sum(axis=1)
    priority_vector = row_sum / len(normalized_matrix)
    priority_vectors.append(priority_vector)

    # Create the DataFrame from the priority scale
    priority_df = pd.DataFrame(priority_vector, index=heroes, columns=[criteria[i]])
    print(f"Skala Prioritas {criteria[i]}:\n")
    print(priority_df)
    print("\n")

Skala Prioritas Health:

                  Health
Miya            0.007952
Balmond         0.008620
Saber           0.008188
Alice           0.008106
Nana            0.007879
Tigreal         0.008475
Alucard         0.008258
Karina          0.008295
Akai            0.008724
Franco          0.008535
Bane            0.008062
Bruno           0.007946
Clint           0.007971
Rafaela         0.007690
Eudora          0.007952
Zilong          0.008472
Fanny           0.007958
Layla           0.007876
Minotaur        0.009007
Lolita          0.008125
Hayabusa        0.007653
Freya           0.008825
Gord            0.007807
Natalia         0.008314
Kagura          0.008053
Chou            0.008532
Sun             0.008689
Alpha           0.008336
Ruby            0.008336
Yi Sun-shin     0.008097
Moskov          0.007735
Johnson         0.008850
Cyclops         0.007942
Estes           0.006997
Hilda           0.008787
Aurora          0.007879
Lapu-Lapu       0.008280
Vexana          0.007816


In [124]:
# Combine the priority scale with each criteria
priority_matrix = np.column_stack(priority_vectors)
priority_matrix_df = pd.DataFrame(priority_matrix, index=heroes, columns=criteria)
print("Matriks Skala Prioritas:\n")
print(priority_matrix_df)

Matriks Skala Prioritas:

                  Health  Health Regen  Attack Speed    Damage  \
Miya            0.007952      0.006351      0.008281  0.007005   
Balmond         0.008620      0.009949      0.008281  0.008336   
Saber           0.008188      0.007409      0.008438  0.008266   
Alice           0.008106      0.007621      0.007813  0.007985   
Nana            0.007879      0.007197      0.008438  0.008055   
Tigreal         0.008475      0.008891      0.008047  0.007845   
Alucard         0.008258      0.008256      0.008750  0.008616   
Karina          0.008295      0.008256      0.008750  0.008476   
Akai            0.008724      0.008891      0.008281  0.008055   
Franco          0.008535      0.009738      0.008047  0.008126   
Bane            0.008062      0.008891      0.008281  0.008196   
Bruno           0.007946      0.006351      0.007891  0.009807   
Clint           0.007971      0.007621      0.008203  0.008055   
Rafaela         0.007690      0.007621      0.0077

In [125]:
# Calculate the ranking
ranking = np.dot(priority_matrix, priority_scale)

# Create a DataFrame with the ranking results
ranking_df = pd.DataFrame(ranking, index=heroes, columns=['Ranking'])

# Sort the DataFrame by the 'Ranking' column in descending order
ranking_df_sorted = ranking_df.sort_values(by='Ranking', ascending=False)

# Print the sorted ranking
pd.set_option('display.max_rows', None)
print("Ranking Hero (sorted):\n")
print(ranking_df_sorted)

Ranking Hero (sorted):

                 Ranking
Hylos           0.010359
Phoveus         0.009288
Gloo            0.009281
Yu Zhong        0.009167
Paquito         0.009127
Terizla         0.009004
Kaja            0.008911
Aldous          0.008892
Franco          0.008844
Lolita          0.008795
Johnson         0.008783
Aulus           0.008751
Silvanna        0.008727
Dyrroth         0.008725
Grock           0.008720
Julian          0.008714
Sun             0.008712
Akai            0.008706
Atlas           0.008697
Khufra          0.008669
Baxia           0.008645
Yin             0.008627
Alpha           0.008580
Ruby            0.008576
Hilda           0.008565
Jawhead         0.008539
Badang          0.008538
Martis          0.008498
Joy             0.008497
Barats          0.008490
Arlott          0.008472
Carmilla        0.008457
Karina          0.008444
Chou            0.008433
Minsitthar      0.008381
Khaleed         0.008376
Balmond         0.008353
Aamon           0.008351
G